iothub/device/src/Transport/ProtocolRoutingDelegatingHandler.cs

```diff
@@ -1,133 +1,109 @@
-            }
-        }
+                await _handlerLock.WaitAsync().ConfigureAwait(false);
 
-        async Task TryOpenPrioritizedTransportsAsync(bool explicitOpen, CancellationToken cancellationToken)
-        {
-            Exception lastException = null;
+                if (!_transportSelectionComplete)
+                {
+                    // Try next protocol if we're still searching.
 
-            // Concrete Device Client creation was deferred. Use prioritized list of transports.
-            foreach (ITransportSettings transportSetting in this.Context.Get<ITransportSettings[]>())
-            {
+                    ITransportSettings[] transportSettingsArray = this.Context.Get<ITransportSettings[]>();
+                    Debug.Assert(transportSettingsArray != null);
 
-                if (Logging.IsEnabled) Logging.Info(
-                    this,
-                    $"Trying {transportSetting?.GetTransportType()}",
-                    $"{nameof(ProtocolRoutingDelegatingHandler)}.{nameof(TryOpenPrioritizedTransportsAsync)}");
+                    // Keep cycling through all transports until we find one that works.
+                    if (_nextTransportIndex >= transportSettingsArray.Length) _nextTransportIndex = 0;
 
-                if (cancellationToken.IsCancellationRequested)
-                {
-                    if (Logging.IsEnabled) Logging.Info(this, $"Cancellation requested for {Logging.GetHashCode(cancellationToken)}.", $"{nameof(ProtocolRoutingDelegatingHandler)}.{nameof(TryOpenPrioritizedTransportsAsync)}");
-                    var tcs = new TaskCompletionSource<bool>();
-                    tcs.SetCanceled();
-                    await tcs.Task.ConfigureAwait(false);
+                    ITransportSettings transportSettings = transportSettingsArray[_nextTransportIndex];
+                    Debug.Assert(transportSettings != null);
+
+                    if (Logging.IsEnabled) Logging.Info(
+                        this,
+                        $"Trying {transportSettings?.GetTransportType()}",
+                        $"{nameof(ProtocolRoutingDelegatingHandler)}.{nameof(OpenAsync)}");
+
+                    // Configure the transportSettings for this context (Important! Within Context, 'ITransportSettings' != 'ITransportSettings[]').
+                    Context.Set<ITransportSettings>(transportSettings);
+                    CreateNewTransportHandler();
+
+                    _nextTransportIndex++;
                 }
 
                 try
                 {
-                    this.Context.Set(transportSetting);
-                    this.InnerHandler = this.ContinuationFactory(this.Context);
-
-                    // Try to open a connection with this transport
-                    await base.OpenAsync(explicitOpen, cancellationToken).ConfigureAwait(false);
+                    await base.OpenAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
@prmathur-microsoft this handler should be removed and we should only retry when the error is `IotHubCommunicationException`.